src/Orleans.Pipeline.Server/OrleansPipeGrain.cs

```diff
@@ -25,25 +29,128 @@ protected Task Notify(TFromServer data)
             }));
     }
 
-    public Task Write(PipeTransferItem<TToServer> item)
+    public override Task OnActivateAsync(CancellationToken cancellationToken)
+    {
+        _heartbeatTimer = this.RegisterGrainTimer(PeriodicallySendHeartbeat, new GrainTimerCreationOptions()
+        {
+            Interleave = true, // allow interleaving so clients can still write data while heartbeat is running.
+            DueTime = HeartbeatPeriod,
+            Period = HeartbeatPeriod
+        });
+
+        return Task.CompletedTask;
+    }
+
+    public override Task OnDeactivateAsync(DeactivationReason reason, CancellationToken cancellationToken)
+    {
+        _heartbeatTimer?.Dispose();
+        _heartbeatTimer = null;
+
+        return Task.CompletedTask;
+    }
+
+    public async Task Write(PipeTransferItem<TToServer> item)
     {
         logger.LogInformation("Received transfer with mode {mode}", item.Mode);
-        return item switch
+
+        var task = item switch
         {
             { Mode: TransferMode.Data, Item: not null } => OnData(item.Item),
             _ => Task.CompletedTask
         };
+
+        await task; // await the task so it can interleave with the heartbeat mechanism
     }
 
     public Task Subscribe(IOrleansPipeObserver<TFromServer> observer)
     {
-        _observers.Subscribe(observer, observer);
+        _manager.Subscribe(observer, observer);
         return Task.CompletedTask;
     }
 
     public Task Unsubscribe(IOrleansPipeObserver<TFromServer> observer)
     {
-        _observers.Unsubscribe(observer);
+        _manager.Unsubscribe(observer);
         return Task.CompletedTask;
     }
+
+    private async Task PeriodicallySendHeartbeat()
+    {
+        if (_manager.Count == 0)
+        {
+            _heartbeatTimer?.Dispose();
+            _heartbeatTimer = null;
+
+            DeactivateOnIdle();
+
+            if (logger.IsEnabled(LogLevel.Information))
+            {
+                logger.LogInformation("PipeGrain with Id = {GrainId} is deactivating because " +
+                    "there are no observers attached anymore.", this.GetPrimaryKeyString());
+            }
+
+            return;
+        }
+
+        await _manager.Notify(_ => Task.CompletedTask); // clear any potentially defunct once, but dont notify serially.
+
+        var healthyObservers = _manager.Observers.Select(x => x.Value); // no need to snapshot, its already one.
+
+        var key = this.GetPrimaryKeyString();
+        var hbTasks = new List<Task<IOrleansPipeObserver<TFromServer>>>();
+
+        foreach (var observer in healthyObservers)
+        {
+            hbTasks.Add(SendHeartbeat(observer, key));
+        }
+
+        _heartbeatCts.CancelAfter(HeartbeatTimeout);
+        await foreach (var hbTask in Task.WhenEach(hbTasks).WithCancellation(_heartbeatCts.Token))
```

Comment:
is the obersver.ReceiveMessage call blocking, or is this timeout just for safety?

(Reason for asking: after this pr i want to get all values out into configuration and documented)

Reply:
just for safety